### PR TITLE
Add scan-hardcoded-strings skill for non-localized UI text inventory

### PR DESCRIPTION
**Pull request by rajbos**
## Summary

- Adds a new on-demand, informational skill, `.github/skills/scan-hardcoded-strings/`, that scans `vscode-extension/src/webview/**/*.ts` and the `get*Html()` methods in `vscode-extension/src/extension.ts` for string/template literals in UI-rendering positions — `.textContent`/`.innerText`/`.innerHTML`/`.title`/`.placeholder` assignments (direct literal or a conditional/ternary RHS, including multi-line with template-literal branches, and quote-aware statement-boundary matching), `aria-label="..."`/`title="..."`/`placeholder="..."` HTML attributes (or the same via `.setAttribute(...)`), `document.createTextNode(...)` literals, text inside `<div>`/`<button>`/`<vscode-button>`/`<label>`/`<h1>`–`<h6>`/`<p>`/`<span>`/`<td>`/`<th>`/`<option>`/`<summary>`/`<caption>`/`<li>`/`<title>`/SVG `<text>` tags (tolerating nested inline and void/structural tags, including a nested `<span>` badge), and the UI-text argument of the shared DOM helpers `el()`/`iconHeading()`/`createButton()` — including a `+`-concatenation or a top-level ternary/`??`/`||` fallback of literal branches, not just one direct literal — that are **not** wrapped in `localize(`, `localizeFormat(`, `t(`, or `vscode.l10n.t(`.
- The script (`scan-hardcoded-strings.js`) is dependency-free plain Node, mirrors the structure/conventions of the existing `check-urls` skill, is informational only (sets `process.exitCode = 0`, never gates, catches operational errors via an injectable `runMain()`), and supports `--json` and `--out <path>` flags. Exports its pure helpers via `module.exports`.
- Comment masking (`/* ... */`/JSDoc blanking) uses a proper stack-based tokenizer — tracking code / string / template-literal / template-interpolation / comment spans — so it correctly handles regex literals (`.replace(/"/g, ...)`), quote-containing UI strings, and *nested* template literals (`` `...${fn(`${x} y`)}...` ``, common in this HTML-templating codebase) without desyncing and silently disabling masking for the rest of a file. Verified zero unmasked comment markers across every file the scanner processes.
- Every balanced-bracket/argument-splitting/statement-boundary scan in the script (matching a call's closing paren, a method's closing brace, splitting call arguments on top-level commas, splitting a `+`-concatenation, a `.title=`-style template-literal assignment, a conditional assignment's terminating `;`) shares one quote/template-aware primitive (`skipQuotedLiteral`) that correctly skips a *nested* template literal or a semicolon inside a quoted branch, so neither a helper call's argument list, a property assignment, nor a ternary statement gets truncated.
- Interpolation-literal recovery (`${...}`) preserves a literal that is a whole ternary branch (`?`/`:`) or the fallback side of a `||` default (e.g. `${msg || '<em>No message</em>'}`), while still excluding a plain quoted literal that's merely an argument to some other call in the same interpolation (e.g. a button id in `${buttonHtml('btn-refresh')}`) or a comparison operand (`typeof x === 'string'`) — but a **template** literal is recovered regardless of position, since it's used to build display text (e.g. `${escapeHtml(`${p.kind} · ${p.model} AIU`)}` still surfaces "AIU"). The same `??`/`||` fallback support extends to helper-call text arguments (e.g. `KIND_LABEL[feature.kind] ?? 'New'`). HTML character references (`&middot;`, `&#8226;`) are stripped before the prose check so an entity name isn't itself mistaken for prose.
- `scan-hardcoded-strings.test.js` — 130 unit tests (`node:test`) covering every helper and detector, the `extension.ts` scope restriction, comment-masking edge cases (regex literals, nested templates), nested-template-aware bracket/argument/assignment/statement-boundary matching, ternary/`??`/`||`-fallback interpolation and helper-argument recovery, HTML entity stripping, the `runMain()` error contract, and CLI smoke tests against a scratch `--out` path.
- `README.md` and `SKILL.md` document what it does, how to run it and its tests, and how to triage its output — including the full webview localization path and correct `npm --prefix vscode-extension run lint:l10n`/`validate:l10n` invocations.
- Per the repo's "Keep Claude Code's Mirrored Agents & Skills in Sync" rule, `SKILL.md` is copied verbatim to `.claude/skills/scan-hardcoded-strings/SKILL.md`. Also added to `.github/skills/README.md`'s Available Skills catalog.
- Ran the script against the current codebase (post-merge with `main`) and committed the resulting `hardcoded-strings-report.md` (repo root) as a visible initial backlog: **1226 candidate strings across 16 files** (largest: `usage/main.ts`, `diagnostics/main.ts`, `logviewer/main.ts`, `efficiency/main.ts`).

This does not fix any of the flagged strings — that's intentionally out of scope for this change, which only builds the scanner and produces the inventory. It's independent of the separate in-flight strict AST-based CI gate work (different files, no overlap).

## Test plan

- [x] `node --check .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js` — syntax OK
- [x] `node --test .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.test.js` — 130/130 passing, including CLI smoke tests
- [x] Wrote a manual whole-codebase check confirming zero unmasked comment markers and identical input/output length across all 36 scanned files — this is what surfaced the regex-literal and nested-template-literal masking bugs
- [x] Spot-checked findings across many files for false positives after each round of fixes
- [x] Confirmed the script sets `process.exitCode = 0` and never crashes with a non-zero exit, even on an operational error
- [x] Confirmed `.claude/skills/scan-hardcoded-strings/SKILL.md` is byte-identical to `.github/skills/scan-hardcoded-strings/SKILL.md`
- [x] Confirmed running the test suite doesn't rewrite the tracked `hardcoded-strings-report.md`
- [x] Addressed ten rounds of Copilot code review across scope, false positives, coverage gaps, Unicode support, quote/bracket/regex/template-literal tokenizer robustness, ternary/`??`/`||`-fallback interpolation recovery, HTML entity handling, and documentation accuracy — the committed report count matches the description after each round

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_0114aUZCxZgtsJxTJ6hxshaj

### CODE REVIEW

File: .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+    } else {
+        lines.push('## Findings by file');
+        lines.push('');
+        for (const { file, findings } of results) {
+            lines.push(`### \`${file}\` (${findings.length})`);
+            lines.push('');
+            lines.push('| Line | Kind | Snippet |');
+            lines.push('|---|---|---|');
+            for (const f of findings) {
+                const escaped = f.snippet.replace(/\|/g, '\\|');
```
[the remark this answers is not in the document]
Comment (rajbos) [reply .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js]: Fixed in 5f4f720: escape backslashes before pipes (and neutralize literal backticks so a snippet can't close its surrounding code span early) in `escapeTableCell`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+    }
+    deduped.sort((a, b) => a.line - b.line);
+    return deduped;
+}
+
+// ── Main ───────────────────────────────────────────────────────────────────
+
+function main() {
+    const files = [...collectTsFiles(WEBVIEW_DIR)];
+    if (fs.existsSync(EXTENSION_FILE)) { files.push(EXTENSION_FILE); }
```
[the remark this answers is not in the document]
Comment (rajbos) [reply .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js]: Fixed in 5d13042: added `extractHtmlMethodRanges()`, which finds the `get*Html(...)` method bodies via brace-balanced parsing, and `scanFile()` now filters findings to those ranges for `extension.ts`. The `Feedback`/panel-title false positives from `maturityHandleShareToIssue` are gone — `extension.ts` now reports 0 findings (those strings are GitHub issue Markdown / a VS Code panel title, correctly out of scope). Covered by new tests in `scan-hardcoded-strings.test.js`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+        }
+    }
+
+    writeMarkdownReport(results, total, files.length);
+    if (!JSON_MODE) {
+        console.log(`\nMarkdown report written to ${path.relative(REPO_ROOT, REPORT_PATH)}`);
+    }
+
+    // Informational script — never fail the build.
+    process.exit(0);
```
[the remark this answers is not in the document]
Comment (rajbos) [reply .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js]: Fixed in 5d13042: replaced `process.exit(0)` with `process.exitCode = 0` so buffered stdout can drain naturally before the process exits.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/skills/scan-hardcoded-strings/README.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  code
+
+The script always exits `0`. Nothing is auto-fixed.
+
+## After Reviewing Findings
+
+1. For a genuine hardcoded string, add a key to
+   `vscode-extension/package.nls.json` (extension-host code, via `t()`) or to
+   `DEFAULT_LOCALIZATION` in `vscode-extension/src/webview/shared/localization.ts`
+   (webview code, via `localize()`), then reference it from the flagged site.
```
[the remark this answers is not in the document]
Comment (rajbos) [reply .github/skills/scan-hardcoded-strings/README.md]: Fixed in 5d13042: the checklist now spells out both steps for webview text — `DEFAULT_LOCALIZATION` (English fallback) **and** adding the key to `getWebviewLocalization()` in `extension.ts` (via `l10n.t('key')`) plus `package.nls.json`/`package.nls.zh-cn.json` — and explains what breaks if the second step is skipped.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/skills/scan-hardcoded-strings/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+| Genuine hardcoded UI prose | Add a localization key and reference it via `localize()` (webview) or `t()` (extension host) |
+| False positive (e.g. unusual formatting the regex misjudged) | No action — this is a line-scan, not an AST parse; use judgement |
+| Text intentionally not localized (e.g. a raw model/tool name) | No action, but consider a code comment if the reason isn't obvious |
+
+## After Finding Hardcoded Strings
+
+1. Add the missing key to `vscode-extension/package.nls.json` (+ the zh-CN
+   file where translated) for extension-host text, or to
+   `DEFAULT_LOCALIZATION` in
+   `vscode-extension/src/webview/shared/localization.ts` for webview text.
```
[the remark this answers is not in the document]
Comment (rajbos) [reply .github/skills/scan-hardcoded-strings/SKILL.md]: Fixed in 5d13042 (same correction as the README): the instructions now cover `getWebviewLocalization()` and the `package.nls*.json` step for webview keys, not just `DEFAULT_LOCALIZATION`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+function isWithinRanges(index, ranges) {
+    return ranges.some(([start, end]) => index >= start && index <= end);
+}
+
+// ── Localization-call stripping ────────────────────────────────────────────
+
+// Matches the opening of a localize(/t(/l10n.t(/vscode.l10n.t( call. `\b` before
+// the bare `t(` alternative keeps it from matching inside identifiers like
+// `getText(` or `format(` (no word boundary exists between "x" and "t" there).
+const CALL_OPEN = /\b(?:vscode\.l10n\.t|l10n\.t|localize|t)\(/;
```
[the remark this answers is not in the document]
Comment (rajbos) [reply .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js]: Fixed in 9267453: `CALL_OPEN` now includes `localizeFormat` alongside `localize`/`t`/`vscode.l10n.t`/`l10n.t`, and the paren-balancer removes the whole call including its extra format args. Added a regression test confirming `localizeFormat('key', n)` inside an interpolation is fully stripped (and that matching "localize" doesn't cause a bad partial match against "localizeFormat(").

---
_Generated by [Claude Code](https://claude.ai/code)_

File: hardcoded-strings-report.md [hunk 7]
```diff
@@ -0,0 +1,1030 @@
...
+This is an informational triage report, not a CI gate. It lists candidate UI
+strings under `vscode-extension/src/webview/**` and the `get*Html()` methods in
+`vscode-extension/src/extension.ts` that are rendered as UI text but are not
+wrapped in `localize()`, `t()`, or `vscode.l10n.t()`. Some entries may be false
+positives (see the skill README for the exclusion rules) — triage each one
+before localizing.
+
+- Files scanned: 36
+- Files with findings: 16
+- Total candidate strings: 933
```
[the remark this answers is not in the document]
Comment (rajbos) [reply hardcoded-strings-report.md]: Regenerated in 9267453 (and again after the follow-up fixes for `<div>`/`title`/`placeholder`/`localizeFormat`/single-token coverage) — the committed `hardcoded-strings-report.md` is now current: 1121 candidates across 16 files. Updated the PR description to match so there's one authoritative count.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js [hunk 8]
```diff
@@ -0,0 +1,515 @@
...
+    'active', 'hidden', 'visible', 'collapsed', 'expanded', 'disabled', 'enabled', 'selected', 'checked',
+    'flex', 'block', 'inline', 'inline-block', 'grid', 'table', 'none', 'contents',
+    'absolute', 'relative', 'fixed', 'sticky', 'static',
+    'left', 'right', 'center', 'top', 'bottom', 'auto', 'inherit', 'initial', 'unset',
+    'bold', 'italic', 'underline', 'uppercase', 'lowercase', 'capitalize',
+    'pointer', 'default', 'wrap', 'nowrap', 'row', 'column',
+    'solid', 'dashed', 'dotted', 'transparent',
+]);
+const URL_RE = /^(https?:\/\/|www\.)/i;
+const LETTER_RUN_RE = /[A-Za-z]{2,}/;
```
[the remark this answers is not in the document]
Comment (rajbos) [reply .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js]: Fixed in dc15c38: `LETTER_RUN_RE` is now `/\p{L}{2,}/u` (Unicode letter property) instead of `[A-Za-z]{2,}`, so non-Latin scripts are no longer exempted. Added a regression test asserting `looksLikeProse('刷新')` and `looksLikeProse('Сохранить')` both return `true`.

---
_Generated by [Claude Code](https://claude.ai/code)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js [hunk 9]
```diff
@@ -0,0 +1,651 @@
...
+
+    return findings;
+}
+
+const HTML_ATTR_NAMES = ['aria-label', 'title', 'placeholder'];
+
+/**
+ * Detect `aria-label="..."` / `title="..."` / `placeholder="..."` HTML
+ * attributes with hardcoded prose. HTML attributes in this codebase are
+ * always written tight (`title="..."`, no spaces around `=`), so requiring
```
[the remark this answers is not in the document]
Comment (rajbos) [reply .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js:485]: Fixed in b9b812d: added `findSetAttributeCalls()`, which flags a literal passed to `.setAttribute('title'|'aria-label'|'placeholder', '...')`. Both cited call sites (`usage/main.ts:6084`, `diagnostics/main.ts:1417`) now appear in the inventory. Also added `findTextNodeCalls()` for `document.createTextNode(...)` while addressing a related finding in the same review round.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.test.js [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+
+// ── CLI smoke tests (the actual executable path) ─────────────────────────────
+//
+// Everything above tests the exported pure helpers directly; these two tests
+// instead invoke the script exactly as a user/CI would, so a regression in
+// main()'s wiring (argv parsing, console output, report writing) is caught
+// even though it isn't itself an exported function.
+
+test('CLI smoke test: normal mode exits 0, scans real files, and writes the report', () => {
+    const result = spawnSync(process.execPath, [SCRIPT_PATH], { encoding: 'utf8' });
```
[the remark this answers is not in the document]
Comment (rajbos) [reply .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.test.js]: Fixed in b9b812d: the report path is now overridable via `--out <path>` (mirroring `scripts/loc-stats.js`'s own convention), and the CLI smoke tests point it at a scratch file in `os.tmpdir()` instead of the tracked repo-root `hardcoded-strings-report.md`. Verified via checksum that running the test suite no longer changes that file.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: hardcoded-strings-report.md [hunk 11]
```diff
@@ -0,0 +1,1337 @@
...
+| Line | Kind | Snippet |
+|---|---|---|
+| 22 | <div> content | `<div style="font-size:16px;color:#b8b8c8;">${reportLabel} &middot; ${escapeHtml(when)}</div>` |
+
+### `vscode-extension/src/webview/shared/formatUtils.ts` (2)
+
+| Line | Kind | Snippet |
+|---|---|---|
+| 135 | <div> content | `<div style="color: var(--text-secondary); font-size: 12px; padding: 8px 0;"> This section couldn't be displayed due to an unexpected error. Other sections are u…` |
+| 224 | <a> content | `<a href="url" target="_blank" rel="noopener noreferrer">text</a> */ export function markdownToHtml(text: string): string { let escaped = escapeHtml(text); escap…` |
```
[the remark this answers is not in the document]
Comment (rajbos) [reply hardcoded-strings-report.md]: You were right that this was a real bug, not just a stale report — thank you for catching it. Root cause (fixed in 8395f69): `maskBlockComments()` didn't understand regex literals. `.replace(/"/g, ...)`-style HTML-escaping regexes are common in this codebase, and the `"` inside `/"/g` was misread as opening an unterminated string, which silently disabled comment masking for the rest of the file — that's exactly how this JSDoc example leaked through as if it were rendered markup. While fixing that I found a second, related bug in the same function (nested template literals closing the outer one at the first inner backtick) and rewrote the whole thing as a proper stack-based tokenizer. Verified zero unmasked comment markers across all 36 scanned files. Regenerated the report — `formatUtils.ts` is now down to its 1 genuine finding.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: hardcoded-strings-report.md [hunk 12]
```diff
@@ -0,0 +1,1320 @@
...
+This is an informational triage report, not a CI gate. It lists candidate UI
+strings under `vscode-extension/src/webview/**` and the `get*Html()` methods in
+`vscode-extension/src/extension.ts` that are rendered as UI text but are not
+wrapped in `localize()`, `localizeFormat()`, `t()`, or `vscode.l10n.t()`. Some entries may be false
+positives (see the skill README for the exclusion rules) — triage each one
+before localizing.
+
+- Files scanned: 36
+- Files with findings: 17
+- Total candidate strings: 1218
```
[the remark this answers is not in the document]
Comment (rajbos) [reply hardcoded-strings-report.md]: Regenerated in fd91c37 after this round's scanner fixes (nested-template bracket matching, ternary helper-call args, `||` fallback recovery) — the committed `hardcoded-strings-report.md` is now 1225 candidates across 17 files, and I've updated the PR description to match so there's one authoritative count.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js [hunk 13]
```diff
@@ -0,0 +1,1089 @@
...
+ * data..."`. Returns `null` if a (non-ternary) expression's pieces aren't
+ * all direct literals, or if a ternary's branches resolve to nothing (a
+ * variable/call there means the text isn't fully knowable statically, so
+ * that piece is skipped rather than guessed at).
+ */
+function extractConcatenatedLiteralText(expr) {
+    const ternary = splitTernary(expr);
+    if (ternary) {
+        const branchTexts = ternary
+            .map((branch) => extractConcatenatedLiteralText(branch))
```
[the remark this answers is not in the document]
Comment (rajbos) [reply .github/skills/scan-hardcoded-strings/scan-hardcoded-strings.js:702]: Fixed in ab258c7: added `splitNullishOrOr()`, which also covers the symmetric `||` fallback, and used it in `extractConcatenatedLiteralText()`. `KIND_LABEL[feature.kind] ?? 'New'` now surfaces "New" in the inventory. Added a regression test.

---
_Generated by [Claude Code](https://claude.ai/code)_
